fuel-block-producer/src/adapters.rs

```diff
@@ -0,0 +1,38 @@
+use crate::ports::TxPool;
+use fuel_core_interfaces::{
+    common::fuel_tx::{
+        CheckedTransaction,
+        ConsensusParameters,
+    },
+    model::BlockHeight,
+    txpool::Sender,
+};
+use std::sync::Arc;
+
+pub struct TxPoolAdapter {
+    pub sender: Sender,
+    pub consensus_params: ConsensusParameters,
+}
+
+#[async_trait::async_trait]
```

Comment:
I had a similar conversation with @freesig about our overuse of async. However, I'm not totally convinced block production shouldn't be async yet. Given that block production has to also call `execute` on the block which may take a few seconds to process, do you think it makes sense to allow the caller of produce to yield while we do the actual execution on a worker pool (like rayon)?

Reply:
I agree that for `BlockExecuter::execute` we want `async`, because, with the metadata, we can parallelize execution.

But the block **production** should be synchronous because it is our source of metadata. We can't parallelize it, so it will consume the thread's resources entirely. We also can't start the next block production while the previous is not finished.

It is one of two heavies operations(the second is block commit). And so it should spawn a separate thread that fully focuses on it.

Reply:
One risk of making this not async, is that to perform block production it will likely need to be spawned onto the blocking pool. The same blocking pool as the executor would be running on. This busy waiting on the spawned block execution task is an inefficient use of threads and is exactly the sort of thing async is designed to avoid.

> We also can't start the next block production while the previous is not finished.

This is also a valid concern, but could also be dealt with by adding a semaphore to this method and leaving it async.

Reply:
Added an async-friendly semaphore to prevent concurrent block production requests.

Reply:
> so it should spawn a separate thread that fully focuses on it.

If this thread spends most of its time dedicated to being blocked on the executor threadpool, why not just make it an async task on a shared runtime?

Reply:
I created an issue to defer this discussion to a future discovery spike: https://github.com/FuelLabs/fuel-core/issues/642

Let's just keep this async for now.